Powers, Roots, and more

Jsoto22 opened the pull request:
This update allows for fractional exponents to be used in the pow method. It also adds public methods for square/cube roots, exponentiation of Euler's constant, Euler's constant as a static property, and factorials.

Jsoto22 committed:
Integer only exponents

Jsoto22 committed:
Removed bitwise operations

- Allows for any size Integer to be used as an exponent

Jsoto22 committed:
Suggested enhancement

Adds 'Euclidean division' Modulus operation and simplifies Modulus Remainder sign logic. Unit tests added.

Jsoto22 committed:
Factional exponents for Powers; Roots added

Implements support for fractional exponents in powers and integer only roots

Jsoto22 committed:
Additional Updates

Adds new features and updates REAMDME.md
Todo: expand unit testing for new features.

Jsoto22 committed:
Public modulusE method

Jsoto22 commented:
@royNiladri Hope I'm not too late on this but had to work out some precision issues. They're mostly resolved, but I plan to maybe add user defined limits on the public facing method for `pow`. I'll try to wrap this up by finishing the unit testing tonight, then it should be ready for review.

Jsoto22 committed:
Fixes and additions

- Fixed typing issues in BigDecimal validate
- Removed deprecated substr functions
- Added wrapper logic functions to compareTo
- Fully implemented 'UNNECESSARY' rounding mode

Jsoto22 commented:
@royNiladri latest commit implements "UNNECESSARY" rounding mode into roundOff method

Added unit test for roots. Should close #126

Looked into #22 and have a basic plan to implement [Locale Aware numbering systems](https://github.com/unicode-org/cldr-json/tree/main/cldr-json/cldr-numbers-modern). I'll open a new branch for that after the next release.

Jsoto22 committed:
Update README.md

royNiladri commented:
Hi @Jsoto22 You have some conflicts on this PR. I have added you as a contributor to the repository.. So you can start working on the repository directly. Create a separate branch, and merge all changes from the release branch.

Jsoto22 committed:
Merge branch 'Rel/2024-04-m1' into development-pow

Jsoto22 commented:
@royNiladri I fixed the conflicts directly. I had left them so you could be aware of what would change in those files. This pr should be good to go now.

Jsoto22 committed:
Added trailing 0s for divide by 0 with precision

should close issue #57

Jsoto22 committed:
Update README.md

Jsoto22 committed:
Minor patches and testing

thecaligarmo commented:
I tried a local version of this, and there's a couple errors. For example: `pow(2, 1.7)` returns the wrong number.

I think the error is in line 108 of `pow.ts` where I'm pretty sure you want `exponent = abs(subtract(exponent, remainder))`.

(Or I did the local version incorrectly, which is highly likely)

Jsoto22 commented:
@thecaligarmo I'll look into it tonight. Funny enough, I finally had some time this weekend to fix a few other things I caught. Will get back to you on what I find.

thecaligarmo commented:
Sounds good =) I think I was having issues with super large numbers too. In this case it was likely me though, but just in case, I was having issues with `pow('2.56e237', '1.7')` I think in this case it was a precision problem, but I wasn't sure where I messed up copying your code over.

Jsoto22 commented:
I was working on getting a worker instance for powers, but having trouble with that, so I stashed the changes and tested the first example you gave against `Math.pow`. The answer was the accurate.

I'll do some more testing with your second example and hopefully I'll push a patch tonight or tomorrow. It may well be a precision issue, but please do note that very large numbers may cause the page threat to stall out. (Working on that)

thecaligarmo commented:
> I was working on getting a worker instance for powers, but having trouble with that, so I stashed the changes and tested the first example you gave against `Math.pow`. The answer was the accurate.
> 
> I'll do some more testing with your second example and hopefully I'll push a patch tonight or tomorrow. It may well be a precision issue, but please do note that very large numbers may cause the page threat to stall out. (Working on that)

Sounds good =) I probably copied it over wrong then 😅  Thanks for double checking.

Jsoto22 commented:
No worries and thank you for bringing it up anyways. I'm about to push a temp patch so you can test it out. I wasn't able to stash all the changes, so there's a bit of clutter, but I commented out anything that'll interfere.

I added a script `npm run start-min` that'll just build and compile. It'll skip the unit testing meant for production. Keep in mind this is a development branch, so definitely not polished, and some of the equations may be off 😂 sometimes it's hard to translate them into code properly.

Known bug: cos function assigns the wrong sign (it's shifted by π/2 radians)

Jsoto22 committed:
Squashed commit of the following:

commit f219a8e41aeaa46e04e429707520363e6c0448c1
Author: Jsoto22 <sotojared22@gmail.com>
Date:   Mon Oct 14 00:02:26 2024 -0400

    Temp patch with updates

commit a1c173549116fafc3520706e83de8b1bcfd1915e
Author: Jsoto22 <sotojared22@gmail.com>
Date:   Wed Apr 24 16:57:16 2024 -0400

    Bug fixes, optimizations, and testing

commit fd49fdc000b5f01a933dd93b22d3c35213c58e27
Author: Jsoto22 <sotojared22@gmail.com>
Date:   Tue Apr 23 21:22:47 2024 -0400

    Completed Trig functions

    - Bug fixes
    - Optimizations added

commit 53de852d1780bf3cf63468dc8607a274cdddd3e7
Author: Jsoto22 <sotojared22@gmail.com>
Date:   Sat Apr 20 22:09:03 2024 -0400

    Log and trig functions

    todo: Unit testing

commit 5830249057f4f29f3cdd57838a124a51c3e23415
Author: Jsoto22 <sotojared22@gmail.com>
Date:   Mon Apr 8 05:31:36 2024 -0400

    Minor patches and testing

Jsoto22 commented:
@thecaligarmo you're all set 😁 don't be afraid test out all the extras and lmk if you find any bugs.

Jsoto22 committed:
added precision to pow

-defaults to 32
-doubles precision for fractional approximation
-fixes early escape in nthroot

Jsoto22 commented:
After some testing, it was definitely an issue with the precision and an error with a condition in the nthroot approximation. The method used can blow up to infinity after too many iterations, but it's a hard condition to test for when starting with a very large number base.

Jsoto22 committed:
fix pow accuracy (needs testing)

- modified addition and multiply algorithms to use BigInt for increased speed
- modified comparison functions to optimize efficiency and speed
- extended LN2 constant to include 256 decimal places

Jsoto22 commented:
@thecaligarmo just pushed an update that should fix the precision and accuracy issue for very large numbers. It needs testing, so I ask that if you get the chance, to try it out.

To that note, next weekend I'll clean things up and add unit testing to ensure everything works consistently.

Again, I want to thank you for bringing this to my attention! I didn't account for devs using such large numbers, and honestly, it's hard to verify the results since there aren't many ways to do so, outside of manually verifying the calculations at each step. I'll reflect this use case in my unit testing.

Jsoto22 commented:
@royNiladri I modified the addition and multiplication algorithms in the latest push. If you have time, please review them and let me know if there's any concerns with the new approach. It utilizes `BigInt` to add/subtract or multiply the numbers, while keeping with the pattern of keeping the value as a string. The performance improvements are very noticeable, especially for the more intensive calculations with many iterations.

The same is for the modifications I made to the comparison functions. After some extensive profiling, many comparisons accounted for almost half of the functions execution time, but they've been reduced by upto 50% in some cases. I'll continue to find ways to optimize function that disproportionately effect execution times.

Jsoto22 committed:
patch

- fixed minimum precision issue for pow

Jsoto22 commented:
@thecaligarmo results for `pow('2.56e234', 1.7)` results in:

`311894661319209531784584771685432980054845454108026560935094138793511856282764212132937109641989824509581914299863249513026196567981483561644769299300195688137407957477891048560246796819728012819992719696448971164755792112594320505515925138472616258234140944165115308165049193523856901321517869777359445828429704867899223436061539642474630316981420626238232323738823154459444203384368290866483083783.56857932038864360262508411685488` with a precision of 32 places after the decimal.

royNiladri commented:
@Jsoto22 I had added some comments on some files. Can you take a look at those and fix those as well? Also, looks like I did not add separate spec files for add & subtract. Can you add test fiels and add some test for those?

Also, once this looks good, add a note in the README section nad do a major version change, documenting the change, and adding performance improvemenet result screenshots. 

Another note, i think we have made a release after this.. check and create a new release line and point this PR to that.

Jsoto22 commented:
Sorry, I'm unable to find the comments you mentioned.

I'll likely squash and merge this back to the general development branch/pr I have drafted, then close this one without merging to main/release.

It may take a few more weeks of testing and development to get ready for production. In that time I'll look into a way to benchmark the new major revision against the old one. I'll also update all specs.

For realistic applications, I would prefer that there's a maximum precision of `2^10` places after the decimal point. I'm going to take some time to recalculate all the constants to match, but I can only verify either the first 100 digits or 10,000 for many of them, meaning I'll have to calculate at least the first 10,000 and truncate after verifying they're correct.

Along with that, I'll refactor some of the formulas to other identities that are less computationally intensive.

As a side note, in relation to issue #138, I suggest migrating from Webpack to ESbuild and tsup.

Jsoto22 commented:
@royNiladri just overhauled the division algorithm to match the methods for addition and multiplication. I'm still doing extensive profiling and finding small ways to speed up the methods and manage memory usage. To this point, the last outlier to improve performance is the increment method when rounding. All other utility functions, such as stripping trailing zeros and comparisons, negligibly effect time at this point. There a few small changes I have to make with testing for thresholds/tolerances and reusing values as to not recalculate the next iteration from the start.

Anyway, I just stress tested calculating the natural logarithm of 2 to 10,000 digits after the decimal point. My computer took 1.2 minutes to return the results, with 20% CPU utilization from other programs. I've verified the results were correct on an online reference for constants calculated into the quadrillions of digits.

The average time for each division in that algorithm took 1.2ms, even at a precision of 10,004 digits after the decimal. I'll do more stress testing and verifications, while also putting together a plan to benchmark the new revision that will be statistically significant for the end user.

Jsoto22 committed:
Updates, fixes, and additions

Updates:

- Addition/Subtract algorithm rewrite
- Multiplication algorithm rewrite
- Division algorithm rewrite
- Rounding performance
- CompareTo performance
- increased precision of constants

Fixes:

- Powers algorithm
-- changes to precision approximation

Additions:

- Statistics functions
-- mean
-- median
-- mode
-- variance
-- standard deviation

- New utility functions
-- min
-- max
-- clamp
-- lerp (linear interpolation)
-- inverse lerp
-- step (quantize)
-- random (generate random number string using cryptographically safe random values)

To do:
- fix bug with sign of result for cosine
- increase precision of PI constant

Jsoto22 committed:
Patch, minor refactoring, added Pi constants, Unit Testing followed

- Fixed bug with rounding
- Converted compareTo from localeCompare to BigInt operators

- Fixed most bugs with decimal placement when dividing
** One test case still fails. Needs further investigation **

*Modified all internal methods to only accept string arguments to increase likelihood of browser compiling code that is called often. It also removes toString calls that weren't needed.*

Todo:
- Unit testing will be expanded to cover new methods

Jsoto22 commented:
@royNiladri Profile based on the most recent push. Yesterday, it was clocking in at around 2.57 seconds and I'm not exactly sure where the gains came from, but I'm personally ecstatic at how performant it is. Could have been from my computer restart earlier. The only other improvements I can think of is to store the values as a BigInt and separate value for the decimal index, so that there's fewer temporary resources used and saves time on creating them. Outside of that, I'm looking into a way to simplify how to get the Nth-Root based on the exponent's mantissa so there's no need to iterate and calculate multiple roots per iteration. Everything should be wrapped up and ready for release by the end of the year.

![pow_result](https://github.com/user-attachments/assets/0e37c090-462c-41c8-91c2-e217d0e8ff60)
![pow_profile](https://github.com/user-attachments/assets/68c88690-05bf-4cdc-9f90-0da851900dd7)

Jsoto22 committed:
Patch

- Fixed precision issues

Jsoto22 committed:
Patch

- fixed pow when exponent decimal digit is '3'
- simplified root 10 temp base in exponent decimal calculation

Jsoto22 committed:
Patch, Unit Testing added

- Fixed bug with `10^n`
- Fixed bug with decimal placement when dividing
- Modified Division and Multiplication Stress Test
- Unit Testing added for Statistic methods
- Some Return and Argument types added

Jsoto22 committed:
Simplified division logic

Jsoto22 committed:
Added dedicated exponential calculation and logarithms testing

Jsoto22 committed:
Removed unused utility functions, testing added

Jsoto22 committed:
Dist

Jsoto22 committed:
Cache factorial results for reuse

testing added

Jsoto22 committed:
Square root optimizations

Jsoto22 committed:
Patch and optimization

Jsoto22 committed:
Added different degrees of precision

H = High(1024), M = Medium(512), L = low(128)

Todo: Continue H/M/L pattern for all constants

** May calculate constants beyond the current High precision dynamically, based on the users needs, then cache the results to guarantee accuracy and efficiency**

Jsoto22 committed:
Optimizations

Jsoto22 committed:
Optimizations (experimental)

Changes to formulas

** Further profiling and testing needed**

Jsoto22 committed:
Pi/2 static const updated

Jsoto22 committed:
Dist

Jsoto22 committed:
Removed tables

Jsoto22 committed:
Expanded constants

Jsoto22 committed:
Division patch

Jsoto22 committed:
Exp patch

Jsoto22 committed:
Inverse sqrt patch

Jsoto22 committed:
Trig patches and testing

Jsoto22 committed:
Dist

Jsoto22 commented:
@royNiladri latest push gets UT coverage over 80%. If you could find the time, can you add tests to big-decimal.spec? Keep in mind I added quite a few stress tests, so it make take up to a couple of minutes to complete. I'll increase coverage for pow.ts in the next week.

Jsoto22 commented:
@royNiladri did some profiling on 10,000 divisions against decimal.js. After about 10 rounds, the new division algorithm I put together, with minor changes, is around 18.5% faster. Just about all other functions I've tested are either considerably faster or within a margin of ±5%.

Jsoto22 commented:
@royNiladri @janeklb @eldereal Screenshot of initial benchmarking against decimal.js using Benchmark,js

![image](https://github.com/user-attachments/assets/88fefd79-2e0c-4a0a-99d8-582eb69b5602)

I'll finish the benchmarks by adding all of the core arithmetic and higher order functions. I'll also include other similar libraries to the running. After so, I'll move into cleaning things up, adding comments, finishing the unit tests (they stand to be more thorough), and then updating the readme. In this, I'll ensure that any conflicts with upstream are resolved.

To finish everything up, I'd like to migrate the project from using webpack over to esbuild and tsup. I'll target ES5 but will try to make sure it works with modern frameworks like Vue 2+. Once #138 is resolved, I'll ask for a review and merge to master, then publish to NPM as a major release.

royNiladri commented:
@Jsoto22 This looks great! There is another library called "big.js", we can try profiling with that as well. Yes, we need to fix the build system. We are getting some issues with Vue 2.x. Let me check if i can take this up.

Jsoto22 commented:
@royNiladri these are initial results. I've since made a robust benchmark system to ensure everything tested is one to one. The precision system we use is different and their equivalent functions do not convert to a string after execution like ours. I've accounted for all this though.

As for big.js, it's harder to do a equal comparison because it lack a lot of the features that decimal.js has. I won't be able to compare the 3 side by side but I can do a separate comparison. That being said, lmk what you think about creating a separate library for a light version of BigDecimal.

Jsoto22 commented:
@royNiladri Dropping in to give an update. I last left off finding that the current implementations, specifically the powers, can lead to an error with BigInt. I found this while working on the benchmarks and testing. It still blows my mind that BigInt has limitations to how large of a number it's willing to calculate.

I've spent this last week thinking through a new approach that splits the BigInt into chunks and started implementing it, and completed addition and subtraction. This would considerably change the library from the ground up, causing a complete rewrite, but will optimize the efficiency of the higher math functions. Internally, it'll store BigInt arrays instead of constantly converting to strings and back. I'm testing the optimal chunk size, but I've used 6 digits for testing and fisrt implementations. This can be changed to larger chunk sizes.

As a side note, I've implemented a NTT-based multiplication algorithm for testing. It's highly efficient for very large values, in the order of 100s of thousands of digits, with a complexity of O(nLog(n)). I've optimized my implementation so it's very fast compared to others found online. The ones I've tried stalls the browser before doing any multiplication, while mine can handle multiplying values with 250 thousand digits within a few seconds. Most of the time spent comes from finding a suitable prime and factorization, then converting the BigInt array into a string at the end. This can be cut back if I use predefined constants, but that limits the max digit length the algorithm can handle. Maybe a limit of 13 trillion digits for the product may be good enough, but that's not arbitrary precision 😅

I'm also looking into an implementation based on Karatsuba using the mentioned chunked value storage, but I'd like to avoid recursion if possible. It's much less complicated than the NTT algorithm, but has a complexity of O(n^1.535). This is great for values in the order of 10s of thousand of digits. I may look into implementing both algorithms, along with a naive multiplication algorithm, and switching between them depending on how large the values are.

As far as division goes, I'm sure I can find an algorithm that works with the chunked method. That'll take some time, but it'll likely follow the same principles as the current algorithm in this PR.

Jsoto22 commented:
Any further updates will be committed into the development branch given the drastic changes needed to get around the limitations of BigInt.
